packages/dcos-integration-test/extra/test_networking.py

```diff
@@ -27,9 +30,123 @@ def _ensure_routable():
+    logging.info('VIP_TEST START {}'.format(teststr))
+    agents = list(dcos_api_session.slaves)
+    # make sure we can reproduce
+    random.seed(vip)
+    random.shuffle(agents)
+    host1 = agents[0]
+    host2 = agents[0]
+    if not samehost:
+        host2 = agents[1]
+    logging.info('host1 is is: {}'.format(host1))
+    logging.info('host2 is is: {}'.format(host2))
+    origin_app, app_uuid = get_test_app_in_docker()
+    origin_app['container']['docker']['portMappings'][0]['labels'] = {'VIP_0': vip}
+    origin_app['container']['docker']['network'] = vipnet
+    if vipnet == 'USER':
+        origin_app['ipAddress'] = {'networkName': 'dcos'}
+    if vipnet == 'HOST':
+        origin_app['cmd'] = '/opt/mesosphere/bin/dcos-shell python '
+        origin_app['cmd'] += '/opt/mesosphere/active/dcos-integration-test/util/python_test_server.py '
+        origin_app['cmd'] += '$PORT0'
+        origin_app['container']['docker']['portMappings'][0]['hostPort'] = 0
+        origin_app['container']['docker']['portMappings'][0]['containerPort'] = 0
+        origin_app['container']['docker']['portMappings'][0]['labels'] = {}
+        origin_app['portDefinitions'] = [{'labels': {'VIP_0': vip}}]
+
+    origin_app['constraints'] = [['hostname', 'CLUSTER', host1]]
+    origin_app['constraints'] = [['hostname', 'CLUSTER', host1]]
+
+    proxy_app = get_test_app_in_docker()[0]
+    proxy_app['container']['docker']['network'] = proxynet
+    proxy_app['constraints'] = [['hostname', 'CLUSTER', host2]]
+    if proxynet == 'USER':
+        proxy_app['ipAddress'] = {'networkName': 'dcos'}
+    if proxynet == 'HOST':
+        proxy_app['cmd'] = '/opt/mesosphere/bin/dcos-shell python '
+        proxy_app['cmd'] += '/opt/mesosphere/active/dcos-integration-test/util/python_test_server.py '
+        proxy_app['cmd'] += '$PORT0'
+        proxy_app['container']['docker']['portMappings'][0]['hostPort'] = 0
+        proxy_app['container']['docker']['portMappings'][0]['containerPort'] = 0
+
+    returned_uuid = None
+
+    with contextlib.ExitStack() as stack:
+        stack.enter_context(dcos_api_session.marathon.deploy_and_cleanup(origin_app))
+        sp = stack.enter_context(dcos_api_session.marathon.deploy_and_cleanup(proxy_app))
+        cmd = '/opt/mesosphere/bin/curl -s -f -m 5 http://{}/test_uuid'.format(vipaddr)
+        returned_uuid = ensure_routable(cmd, sp)()
+        logging.info('returned_uuid is: {}'.format(returned_uuid))
+    assert(returned_uuid is not None)
+    assert(returned_uuid['test_uuid'] == app_uuid)
+    logging.info('VIP_TEST PASSED {}'.format(teststr))
+    return app_uuid
+
+
+@pytest.mark.skipif(not lb_enabled(), reason='Load Balancer disabled')
+def test_vip(dcos_api_session):
+    """Test every permutation of VIP
+    """
+    num = 0
+    addrs = [['1.1.1.{}:{}', '1.1.1.{}:{}'],
+             ['/namedvip{}:{}', 'namedvip{}.marathon.l4lb.thisdcos.directory:{}']]
```

Comment:
Ah, yes. Of course.

The idea of using a fixture to start a thread pool like you describe won't really work within the pytest framework because what you would end up having (if I understand this correctly) would be:
```
@pytest.fixture
def my_thread_control():
    # generate all your threads and start them
    yield
    # join your threads and check them
```
and all of your asserting (i.e. the test cases) will be done after joining threads. Anything you define in a test case won't be able to get back into a fixture.

However, I think as we get more and more tests, this issue of parallelization will begin concern more people. I am thinking of building a plugin where you could mark tests like so:
```
@parallel(scope='networking_test')
def test_my_network(some_setup_fixture):
    pass

@parallel(scope='networking_test')
def test_my_network_2(some_setup_fixture):
    pass
```
this would require changing up the runtest protocol, but pytest is really open about this, so I dont think it will be that difficult.

I think for now we should probably just stick to making the huge lot of threads in a single test as you do now, so to keep this focused on the test instead of pytest madness